File: JavaLox/src/com/craftinginterpreters/lox/Lox.java
package com.craftinginterpreters.lox;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import org.apache.hc.client5.http.classic.methods.HttpGet;
import org.apache.hc.client5.http.classic.methods.HttpPost;
import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.core5.http.HttpEntity;
import org.apache.hc.core5.http.NameValuePair;
import org.apache.hc.core5.http.ParseException;
import org.apache.hc.core5.http.io.entity.EntityUtils;
import org.apache.hc.core5.http.message.BasicNameValuePair;
import org.apache.hc.core5.net.URIBuilder;

/**
 * An implementation of the "Lox" embedded scripting language. This work was
 * created from tutorials by the original author, Robert Nystrom. Source URL:
 * https://craftinginterpreters.com/
 *
 * @author Victor G. Brusca, Carlo Bruscani, based on the work of Robert
 * Nystrom, Copyright Middlemind Games 05/03/2024
 */
public class Lox {

   public enum HttpCallType {
      GET_BLANK,
      GET_QUERY_PARAMS,
      POST
   }

   private static final Interpreter interpreter = new Interpreter();
   static boolean hadError = false;
   static boolean hadRuntimeError = false;
   static String lastError = null;

   static String globalsFile = null;
   static String globalsScript = null;
   static boolean hasGlobalsFile = false;
   static boolean hasGlobalsScript = false;
   static String lastFile = null;
   static String lastLine = null;

   /**
    * The static main entry point to the program. Expects a valid Lox script
    * file as an argument or no arguments if you want to start a REPL session.
    *
    * @param args
    * @throws IOException
    */
   public static void main(String[] args) throws IOException {
      hasGlobalsFile = false;
      hasGlobalsScript = false;
      for (int i = 0; i < args.length; i++) {
         if (args[i] != null && args[i].equals("-gf")) {
            if (i + 1 < args.length) {
               globalsFile = args[i + 1];
               hasGlobalsFile = true;
            }
         } else if (args[i] != null && args[i].equals("-gs")) {
            if (i + 1 < args.length) {
               globalsScript = args[i + 1];
               hasGlobalsScript = true;
            }
         }
      }

      if (hasGlobalsFile) {
         Lox.log(0, "Found global file to import: " + globalsFile);
         runFile(globalsFile);
      }

      if (hasGlobalsScript) {
         Lox.log(0, "Found global script to import: " + globalsScript);
         run(globalsScript);
      }

      if (args.length >= 1 && args[0] != null && args[0].toLowerCase().equals("-p")) {
         runPrompt();
      } else if (args.length >= 2 && args[0] != null && args[0].toLowerCase().equals("-f") && args[1] != null) {
         runFile(args[1]);
      } else if (args.length >= 2 && args[0] != null && args[0].toLowerCase().equals("-s") && args[1] != null) {
         run(args[1]);
      } else {
         System.out.println("Usage: JavaLox ([-f script file] | [-s script string] | [-p REPL]) & [-gf script file] [-gs script string]");
         System.exit(64);
      }
   }

   public static String getBlank(String url) {
      String resultContent = null;
      HttpGet httpGet = new HttpGet(url);
      try (CloseableHttpClient httpclient = HttpClients.createDefault()) {
         try (CloseableHttpResponse response = httpclient.execute(httpGet)) {
            // Get status code
            System.out.println(response.getVersion()); // HTTP/1.1
            System.out.println(response.getCode()); // 200
            System.out.println(response.getReasonPhrase()); // OK
            HttpEntity entity = response.getEntity();
            // Get response information
            resultContent = EntityUtils.toString(entity);
         }
      } catch (IOException | ParseException e) {
         e.printStackTrace();
      }
      return resultContent;
   }

   public static String getQueryParams(String url, List<NameValuePair> nvps) {
      String resultContent = null;
      HttpGet httpGet = new HttpGet(url);

      try {
         URI uri = new URIBuilder(new URI(url)).addParameters(nvps).build();
         httpGet.setUri(uri);
      } catch (URISyntaxException e) {
         throw new RuntimeException(e);
      }

      try (CloseableHttpClient httpclient = HttpClients.createDefault()) {
         try (CloseableHttpResponse response = httpclient.execute(httpGet)) {
            System.out.println(response.getVersion()); // HTTP/1.1
            System.out.println(response.getCode()); // 200
            System.out.println(response.getReasonPhrase()); // OK
            HttpEntity entity = response.getEntity();
            resultContent = EntityUtils.toString(entity);
         }
      } catch (IOException | ParseException e) {
         e.printStackTrace();
      }
      return resultContent;
   }

   public static String post(String url, List<NameValuePair> nvps) {
      String result = null;
      HttpPost httpPost = new HttpPost(url);
      httpPost.setEntity(new UrlEncodedFormEntity(nvps));
      try (CloseableHttpClient httpclient = HttpClients.createDefault()) {
         try (CloseableHttpResponse response = httpclient.execute(httpPost)) {
            System.out.println(response.getVersion()); // HTTP/1.1
            System.out.println(response.getCode()); // 200
            System.out.println(response.getReasonPhrase()); // OK

            HttpEntity entity = response.getEntity();
            result = EntityUtils.toString(entity);
            // Ensure that the stream is fully consumed
            EntityUtils.consume(entity);
         }
      } catch (IOException | ParseException e) {
         e.printStackTrace();
      }
      return result;
   }

   private static String runUrl(String url, HttpCallType callType, List<NameValuePair> data) {
      if (callType == HttpCallType.GET_BLANK) {
         return getBlank(url);
      } else if (callType == HttpCallType.GET_QUERY_PARAMS) {
         return getQueryParams(url, data);
      } else if (callType == HttpCallType.POST) {
         return post(url, data);
      } else {
         return null;
      }
   }

   private static void runFile(String path) throws IOException {
      lastFile = path;
      byte[] bytes = Files.readAllBytes(Paths.get(path));
      run(new String(bytes, Charset.defaultCharset()));

      // Indicate an error in the exit code.
      if (hadError) {
         System.exit(64);
      }
      if (hadRuntimeError) {
         System.exit(70);
      }
   }

   private static void runPrompt() throws IOException {
      InputStreamReader input = new InputStreamReader(System.in);
      BufferedReader reader = new BufferedReader(input);

      //Control-D to exit
      for (;;) {
         System.out.println("> ");
         String line = reader.readLine();
         if (line == null || line.toLowerCase().equals("exit") || line.toLowerCase().equals("bye")) {
            break;
         }

         try {
            run(line);
         } catch (Exception e) {
            //do nothing
         }

         hadError = false;
         hadRuntimeError = false;
      }
   }

   private static void run(String source) {
      lastLine = source;
      Scanner scanner = new Scanner(source);
      List<Token> tokens = scanner.scanTokens();

      Parser parser = new Parser(tokens);
      List<Stmt> statements = parser.parse();

      // Stop if there was a syntax error.
      if (hadError) {
         return;
      }

      Resolver resolver = new Resolver(interpreter);
      resolver.resolve(statements);

      // Stop if there was a resolution error.
      if (hadError) {
         return;
      }

      interpreter.interpret(statements);
   }

   static void log(int line, String message) {
      System.out.println("[line " + line + "] Log: " + message);
   }

   static void error(int line, String message) {
      report(line, "", message);
   }

   private static void report(int line, String where, String message) {
      lastError = "[line " + line + "] Error" + where + ": " + message;
      System.err.println(lastError);
      hadError = true;
   }

   static void error(Token token, String message) {
      if (token.type == TokenType.EOF) {
         report(token.line, " at end", message);
      } else {
         report(token.line, " at '" + token.lexeme + "'", message);
      }
   }

   static void runtimeError(RuntimeError error) {
      String msg = error.getMessage() + "\n[line " + error.token.line + "]";
      System.err.println(msg);
      hadRuntimeError = true;
   }
}

File: CsLox/com/craftinginterpreters/lox/Lox.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Security.Principal;
using System.Text;
using System.Threading.Tasks;
using static System.Net.WebRequestMethods;
using System.Net.Http;
using System.Web;
using System.Security.Policy;

namespace com.craftinginterpreters.lox
{
    /// <summary>
    /// An implementation of the "Lox" embedded scripting language.
    /// This work was created from tutorials by the original author, Robert Nystrom.
    /// Source URL: https://craftinginterpreters.com/ 
    /// 
    /// @author Victor G.Brusca, Carlo Bruscani, Copyright Middlemind Games 05/03/2024
    /// </summary>
    public class Lox
    {
        public enum UrlConnType
        {
            GET_BLANK,
            POST,
            GET_QUERY_PARAMS
        }

        private static readonly Interpreter interpreter = new Interpreter();
        internal static bool hadError = false;
        internal static bool hadRuntimeError = false;
        internal static String lastError = null;

        internal static String globalsFile = null;
        internal static String globalsScript = null;
        internal static String globalsUrl = null;
        internal static bool hasGlobalsFile = false;
        internal static bool hasGlobalsScript = false;
        internal static bool hasGlobalsUrl = false;
        internal static String lastFile = null;
        internal static String lastLine = null;
        internal static String lastUrl = null;

        internal static bool hasReturnUrl = false;
        internal static String returnUrl = null;
        internal static String returnUrlGlobalVarName = null;

        /// <summary>
        /// The static main entry point to the program.
        /// Expects a valid Lox script file as an argument or no arguments if you want to start a REPL session.
        /// </summary>
        /// <param name="args"></param>
        public static void Main(String[] args)
        {
            /*
            String url = "";
            List<NameValuePair> data = new List<NameValuePair>();
            data.Add(new BasicNameValuePair("answer", "hello"));

            url = "http://localhost:5223/getGlobal";
            Lox.log(0, "Calling getBlank URL = '" + url + "', Response = \n" + getBlank(url));

            url = "https://localhost:7109/getGlobal";
            Lox.log(0, "Calling getBlank URL = '" + url + "', Response = \n" +  getBlank(url));

            url = "http://localhost:5223/getGlobal";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "https://localhost:7109/getGlobal";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "http://localhost:5223/getGlobal";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            url = "https://localhost:7109/getGlobal";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            url = "http://localhost:5223/getScript";
            Lox.log(0, "Calling getBlank URL = '" + url + "', Response = \n" + getBlank(url));

            url = "https://localhost:7109/getScript";
            Lox.log(0, "Calling getBlank URL = '" + url + "', Response = \n" + getBlank(url));

            url = "http://localhost:5223/getScript";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "https://localhost:7109/getScript";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "http://localhost:5223/getScript";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            url = "https://localhost:7109/getScript";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            url = "http://localhost:5223/setAnswer";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "https://localhost:7109/setAnswer";
            Lox.log(0, "Calling getQueryParams URL = '" + url + "', Response = \n" + getQueryParams(url, data));

            url = "http://localhost:5223/setAnswer";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            url = "https://localhost:7109/setAnswer";
            Lox.log(0, "Calling post URL = '" + url + "', Response = \n" + post(url, data));

            return;
            */

            hasGlobalsFile = false;
            hasGlobalsScript = false;
            for (int i = 0; i < args.Length; i++)
            {
                if (args[i] != null && args[i].Equals("-gf"))
                {
                    if (i + 1 < args.Length)
                    {
                        globalsFile = args[i + 1];
                        hasGlobalsFile = true;
                    }
                }
                else if (args[i] != null && args[i].Equals("-gs"))
                {
                    if (i + 1 < args.Length)
                    {
                        globalsScript = args[i + 1];
                        hasGlobalsScript = true;
                    }
                }
                else if (args[i] != null && args[i].Equals("-gu"))
                {
                    if (i + 1 < args.Length)
                    {
                        globalsUrl = args[i + 1];
                        hasGlobalsUrl = true;
                    }
                }
                else if (args[i] != null && args[i].Equals("-ru"))
                {
                    if (i + 1 < args.Length)
                    {
                        returnUrl = args[i + 1];
                        hasReturnUrl = true;
                    }
                }
                else if (args[i] != null && args[i].Equals("-gv"))
                {
                    if (i + 1 < args.Length)
                    {
                        returnUrlGlobalVarName = args[i + 1];
                    }
                }
            }

            if (hasGlobalsFile)
            {
                Lox.log(0, "Found global file to import: " + globalsFile);
                runFile(globalsFile);
            }

            if (hasGlobalsScript)
            {
                Lox.log(0, "Found global script to import: " + globalsScript);
                run(globalsScript);
            }

            if(hasGlobalsUrl)
            {
                Lox.log(0, "Found global script to import: " + globalsUrl);
                runUrl(globalsUrl, UrlConnType.GET_BLANK, null);
            }

            if (args.Length >= 1 && args[0] != null && args[0].ToLower().Equals("-p"))
            {
                runPrompt();
            }
            else if (args.Length >= 2 && args[0] != null && args[0].ToLower().Equals("-f") && args[1] != null)
            {
                runFile(args[1]);
            }
            else if (args.Length >= 2 && args[0] != null && args[0].ToLower().Equals("-s") && args[1] != null)
            {
                run(args[1]);
            }
            else if (args.Length >= 2 && args[0] != null && args[0].ToLower().Equals("-u") && args[1] != null)
            {
                runUrl(args[1], UrlConnType.GET_BLANK, null);
            }
            else
            {
                /*
                    -f "C:\FILES\OIT_LAPTOP_BACKUP\DOCUMENTS\GitHub\LoxLang\test.lox" 
                    -gs "var GBL_BASE_NAT_LOG = 2.71828;" 
                    -gf "C:\FILES\OIT_LAPTOP_BACKUP\DOCUMENTS\GitHub\LoxLang\globals.lox"
                */
                // -gs "var GBL_BASE_NAT_LOG = 2.71828;" -gf "C:\FILES\OIT_LAPTOP_BACKUP\DOCUMENTS\GitHub\LoxLang\globals.lox" -gu https://localhost:7109/getGlobal -u https://localhost:7109/getScript -ru https://localhost:7109/setAnswer -gv urlGlobal
                //-gu https://localhost:7109/getGlobal -u https://localhost:7109/getScript -ru https://localhost:7109/setAnswer -gv urlGlobal
                String str = "Usage: CsLox  ([-f script file] | [-s script string] | [-u script url] | [-p REPL]) & [-gf script file] [-gs script string] [-gu script url] [-ru url] [-gv global variable name to respond with]";
                System.Diagnostics.Debug.WriteLine(str);
                System.Console.Out.WriteLine(str);
                System.Environment.Exit(64);
            }

            if(hasReturnUrl)
            {
                Dictionary<String, Object> values = interpreter.globals.getValues();
                if (values.ContainsKey(returnUrlGlobalVarName))
                {
                    String value = (values[returnUrlGlobalVarName] + "");
                    List<NameValuePair> data = new List<NameValuePair>();
                    data.Add(new BasicNameValuePair("answer", value));
                    data.Add(new BasicNameValuePair("variableName", returnUrlGlobalVarName));
                    Lox.log(0, "Return URL Response: '" + getQueryParams(returnUrl, data) + "'");
                }
            }
        }

        private static String getBlank(String url)
        {
            Task<String> ret = getBlankWrkrAsync(url);
            return ret.Result;
        }

        private static async Task<String> getBlankWrkrAsync(String url)
        {
            HttpClient client = new HttpClient();
            return await client.GetStringAsync(url);
        }

        private static String getQueryParams(String url, List<NameValuePair> data)
        {
            Task<String> ret = getQueryParamsWrkrAsync(url, data);
            return ret.Result;
        }

        private static async Task<String> getQueryParamsWrkrAsync(String url, List<NameValuePair> data)
        {
            HttpClient client = new HttpClient();
            String newUrl = url;
            if (data != null && data.Count > 0)
            {
                int len = data.Count;
                int cnt = 0;
                newUrl += "?";
                foreach (BasicNameValuePair nvp in data)
                {
                    newUrl += HttpUtility.UrlEncode(nvp.getName()) + "=" + HttpUtility.UrlEncode(nvp.getValue());
                    if(cnt < len - 1)
                    {
                        newUrl += "&";
                    }
                    cnt++;
                }
            }
            return await client.GetStringAsync(newUrl);
        }

        private static String post(String url, List<NameValuePair> data)
        {
            Task<String> ret = postWrkrAsync(url, data);
            return ret.Result;
        }

        private static async Task<String> postWrkrAsync(String url, List<NameValuePair> data)
        {
            HttpClient client = new HttpClient();
            Dictionary<String, String> values = new Dictionary<String, String>();
            if(data != null && data.Count > 0)
            {
                foreach (BasicNameValuePair nvp in data)
                {
                    values.Add(nvp.getName(), nvp.getValue());
                }
            }
            FormUrlEncodedContent content = new FormUrlEncodedContent(values);
            HttpResponseMessage response = await client.PostAsync(url, content);
            Lox.log(0, "ResponseCode: " + response.IsSuccessStatusCode);
            return await response.Content.ReadAsStringAsync();
        }

        private static void runUrl(String url, UrlConnType connType, List<NameValuePair> data)
        {
            String res = "";
            lastUrl = url;
            if (connType == UrlConnType.GET_BLANK)
            {
                res = getBlank(url);
            }
            else if (connType == UrlConnType.GET_QUERY_PARAMS)
            {
                res = getQueryParams(url, data);
            }
            Lox.log(0, "Run URL Response: '" + res + "'");

            JsonScript value = Newtonsoft.Json.JsonConvert.DeserializeObject<JsonScript>(res);
            if(value != null && value.script != "")
            {
                run(value.script);
            }
        }

        private static void runFile(String path)
        {
            lastFile = path;
            byte[] bytes = System.IO.File.ReadAllBytes(System.IO.Path.GetFullPath(path));
            //run(StringFromBytesOptimzed(bytes));              //internal ASCII alternative
            //run(System.Text.Encoding.UTF8.GetString(bytes));  //UTF8 charset alternative
            run(System.Text.Encoding.Default.GetString(bytes)); //default charset

            if (hadError) System.Environment.Exit(65);
            if (hadRuntimeError) System.Environment.Exit(70);
        }

        private static void runPrompt()
        {
            //Control-D to exit
            for(; ;)
            {
                Console.Out.Write("> ");
                String line = Console.ReadLine();
                if (line == null || line.ToLower().Equals("exit") || line.ToLower().Equals("bye")) break;

                try
                {
                    run(line);
                } 
                catch (Exception e) {}

                hadError = false;
                hadRuntimeError = false;
            }
        }

        private static void run(String source)
        {
            lastLine = source;
            Scanner scanner = new Scanner(source);
            List<Token> tokens = scanner.scanTokens();
            Parser parser = new Parser(tokens);
            List<Stmt> statements = parser.parse();

            // Stop if there was a syntax error.
            if (hadError) return;

            Resolver resolver = new Resolver(interpreter);
            resolver.resolve(statements);

            // Stop if there was a resolution error.
            if (hadError) return;

            interpreter.interpret(statements);
        }

        internal static void log(int line, String message)
        {
            String str = "[line " + line + "] Log: " + message;
            System.Diagnostics.Debug.WriteLine(str);
            System.Console.Out.WriteLine(str);
        }

        internal static void error(int line, String message)
        {
            report(line, "", message);
        }

        private static void report(int line, String where, String message)
        {
            lastError = "[line " + line + "] Error" + where + ": " + message;
            System.Diagnostics.Debug.WriteLine(lastError);
            System.Console.Error.WriteLine(lastError);
            hadError = true;
        }

        internal static void error(Token token, String message)
        {
            if (token.type == TokenType.EOF)
            {
                report(token.line, " at end", message);
            }
            else
            {
                report(token.line, " at '" + token.lexeme + "'", message);
            }
        }

        /// <summary>
        /// A helper function that converts a byte array to a string using char casting and a string builder.
        /// </summary>
        /// <param name="b"></param>
        /// <returns></returns>
        public static string StringFromBytesOptimzed(byte[] b)
        {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < b.Length; i++)
            {
                sb.Append((char)b[i]);
            }
            return sb.ToString();
        }

        internal static void runtimeError(RuntimeError error)
        {
            String msg = error.Message + "\n[line " + error.token.line + "]";
            System.Diagnostics.Debug.WriteLine(msg);
            System.Console.Error.WriteLine(msg);
            hadRuntimeError = true;
        }
    }
}
